src/SINDYPI_fit.jl

```diff
@@ -0,0 +1,457 @@
+                basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l]))
+            end
+            end
+        end
+    end
+end
+
+if npoly >= 5
+    for i in 1:nvar
+        for j in i:nvar
+            for k in j:nvar
+                for l in k:nvar
+                    for m in l:nvar
+            basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l] .* X[:,m]))
+        end
+        end
+    end
+end
+end
+end
+
+if npoly >= 6
+    for i in 1:nvar
+        for j in i:nvar
+            for k in j:nvar
+                for l in k:nvar
+                    for m in l:nvar
+                        for n in m:nvar
+            basis =  hcat(basis, (X[:,i].* X[:, j] .* X[:,k] .* X[:,l] .* X[:,m] .* X[:,n]))
+        end
+        end
+    end
+end
+end
+end
+end
+
+if ntrig >=1
+    basis = hcat(basis, sin.(X))
+    basis = hcat(basis, cos.(X))
+end
+
+if ntrig >=2
+    basis = hcat(basis, cos.(X) .* cos.(X))
+end
+
+if ntrig >=3
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* sin.(X[:, j]) .* cos.(X[:, j]) ))
+        end
+    end
+end
+
+if ntrig >=4
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* sin.(X[:, j])))
+        end
+    end
+
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (X[:, i] .* X[:, i] .* cos.(X[:, j])))
+        end
+    end
+
+    for i in 1:nvar
+        for j in 1:nvar
+            basis = hcat(basis, (sin.(X[:, i]) .* cos.(X[:, j]) ))
+        end
+    end
+end
+
+    pin = size(basis,2)
+
+    basis_noderivative = deepcopy(basis[:, 1:pin])
+
+    basis_deep = zeros(size(basis,1), 2*size(basis,2))
+
+    basis_deep[:, 1:size(basis,2)] = basis_noderivative
+
+    for k=1:pin
+        basis_deep[:,pin + k] = Y[:, w].* basis_noderivative[:, k]
+    end
+
+    Right_train = zeros(size(basis_deep, 1), size(basis_deep, 2)-1)
+
+    for i = 1:size(basis_noderivative,2)
+        Right_train[:, i] = basis_deep[:,i]
+    end
+
+    for i = size(basis_noderivative,2)+1:size(basis_deep, 2)-1
+        Right_train[:, i] = basis_deep[:,i+1]
+    end
+
+    Left_train = Y[:, w]
+
+    normLib = zeros(1, size(Right_train,2))
+
+    Theta = deepcopy(Right_train)
+
+        for norm_k=1:size(Theta,2)
+                normLib[norm_k] = norm(Theta[:,norm_k]);
+                Theta[:,norm_k] = Theta[:,norm_k]/normLib[norm_k];
+        end
+
+     Epsilon = zeros(size(Theta,2),1)
+     ldiv!(Epsilon, qr(Theta), Left_train)
+
+    for i in 1:maxiter
+        index_remove = abs.(Epsilon) .<= lambda
+        Epsilon[index_remove] .= zero(eltype(Epsilon))
+
+        for j in 1:1
+            index_retain = @. !index_remove[:,j]
+            Epsilon[index_retain, j] = Theta[:, index_retain] \ Left_train[:,j]
+        end
+    end
+
+        for norm_k=1:length(Epsilon)
+            Epsilon[norm_k,:] = Epsilon[norm_k,:]/normLib[norm_k];
+        end
+
+  Xi[w] = Epsilon
+
+  nvar = length(s)

```

Comment:
This is the backtranslation from epsilon to string, which is not needed.

You basically just provide a method for the computation of epsilon / Xi in place for a sparse optimiser.

So everything else is not needed ( if used within DataDrivenDiffEq ).

I started with a similar approach in my first implementations of the algorithm, but switched to ModelingToolkit since it was way easier to derive everything I needed.